Add Pirate Islands map, re-add vision

Adimote opened the pull request:
Adds the new arena, and updates the API to the new robot API.

f84a624 and 661565b are just getting rid of evidence of previous games,

Please focus on e4c0e9a as it has the most changes.

Adimote committed:
improve the robot looks

Adimote committed:
remove old game yaml files

Adimote committed:
add pirate islands game

Adimote committed:
Modify API to fit the SB2018 API

Robot:
- Remove smallpeice hacks 'MockedRobot' object, integrate it
  into sim_robot
- Remove remains of other arena code
- Add a 'COAST' functionality
- Update demo code to use vision.

Tokens & Markers:
- Update tokens to remove their 'info' 'res' 'centre' and
  'timestamp' properties, add: 'polar', 'cartesian'(unsupported)
  'pixel_centre'(unsupported),

pirate_islands.py:

- move marker ID constants into game_specific.py (same as any other
  game_specific.py)
- Remove re-definition of Token in pirate_islands.py

PeterJCLaw commented on sb/robot/arenas/pirate_islands.py:
```diff
@@ -0,0 +1,254 @@
+        return self._body.angle
+
+    @heading.setter
+    def heading(self, _new_heading):
```
Why does this parameter start with an underscore?

Adimote replied:
fix was bundled in https://github.com/sourcebots/simulator/pull/1/commits/67781c87509a76f0d216df6b92dd998c87e6aa32

RealOrangeOne replied:
This is still an issue, looks like you fixed it in a different place though 👍

Adimote replied:
actually fixed in e78ccd1b3940dfc3ca2c946144bc0633f91d2d97

PeterJCLaw commented on sb/robot/arenas/pirate_islands.py:
```diff
@@ -0,0 +1,254 @@
+            return  # Slight hack: deal with the initial setting from the constructor
+        self._body.angle = _new_heading
+
+    def __init__(self, arena):
```
Please could we have this before the properties?

Adimote replied:
Fixed in https://github.com/sourcebots/simulator/pull/1/commits/94ac0a71b41171d8afd23951b734184be9ffef07

PeterJCLaw commented on sb/robot/arenas/pirate_islands.py:
```diff
@@ -0,0 +1,254 @@
+            return  # Slight hack: deal with the initial setting from the constructor
+        self._body.angle = _new_heading
+
+    def __init__(self, arena, number, damping):
```
Please could we have the `__init__` method before the properties?

PeterJCLaw commented on sb/robot/arenas/pirate_islands.py:
```diff
@@ -0,0 +1,254 @@
+    @location.setter
+    def location(self, new_pos):
+        if self._body is None:
+            return  # Slight hack: deal with the initial setting from the constructor
```
I'm not sure I follow -- we assign `self._body` in `__init__` before we call the `super().__init__()` -- how does the initial setting you're worried about here happen?

Future occurrences not commented.

RealOrangeOne commented on sb/robot/sim_robot.py:
```diff
@@ -36,21 +41,157 @@ def power(self, value):
+                is_wall_marker=lambda: is_wall_marker(mid),
+                is_token_marker=lambda: is_token_marker(mid),
+                polar=polar_coord,
+                pixel_centre="NOT SUPPORTED",
```
I think it would be better to just not have these functions at all rather than this hacky string

Adimote replied:
fixed in https://github.com/sourcebots/simulator/pull/1/commits/4983529b7550dbb68d788d3559925a916e534ec0

RealOrangeOne commented on sb/robot/sim_robot.py:
```diff
@@ -137,10 +278,30 @@ def __init__(self, simulator):
+    __repr__ = __str__
+
+    @property
+    def motor_boards(self):
```
This won't work, because we use a custom dictionary for this. 

You also seem to be missing the single-index properties

PeterJCLaw replied:
Perhaps just drop the plural properties and document that they aren't present in the simulator? (at least for now)

Adimote replied:
I've copied in the `BoardList` class in and done it properly at https://github.com/sourcebots/simulator/pull/1/commits/cef843ef34383f6db7d48174f9f220bbb170b8ff

RealOrangeOne commented on sb/robot/vision.py:
```diff
@@ -0,0 +1,20 @@
+
+# Marker class
+MarkerBase = namedtuple("Marker",
+                        "id size polar pixel_centre cartesian is_wall_marker is_token_marker")
```
I didnt know you could specify this as a string, but I think it's _far_ nicer as a list

PeterJCLaw replied:
+1; note that the second argument can also be a tuple, which is slightly cheaper than a list.

Adimote replied:
The correct way of doing this (at least for python 3.5) is using the `NamedTuple` class. I've done that in https://github.com/sourcebots/simulator/pull/1/commits/df207e06ee5baad65b5abb7139c79dbe616407b6

RealOrangeOne commented on sb/robot/vision.py:
```diff
@@ -0,0 +1,20 @@
+
+
+class Marker(MarkerBase):
+    def __new__(cls, *args, **kwargs):
```
This function call is unnecessary

Adimote replied:
Fixed in https://github.com/sourcebots/simulator/pull/1/commits/afb03723d0c28a26ec66a6f5ed6fa5c3c646f7b6

RealOrangeOne commented on test.py:
```diff
@@ -1,27 +1,49 @@
+from sb.robot import *
```
I always found when running the code that the actual import wasnt necessary, is that not correct anymore?

I also think importing `*` is a terrible anti-pattern that causes so many shadowing issues. Can we be explicit here

PeterJCLaw replied:
Unfortunately this is a bit a hack around the way that the simulator works. What it actually does is place a `Robot` class into the `exec` which runs the user code. (This import isn't actually doing anything).

From what I recall, if you change this import to be precise, then the simulator doesn't work. (I can't remember the specific failure mode).

kierdavis replied:
That sounds slightly horrific. Maybe this is worth having another look into - at the least it'd be nice to document what this line is *actually* doing as opposed to what it appears to do.

Adimote replied:
Not really, as the `test.py` code is intended for teams to copy and use as a basis for their own robots. I think it's best for them to just not know that this line does nothing.

RealOrangeOne replied:
If this import isn't doing anything, we should remove it. I think encouraging people to use this code as a basis is probably a bad idea, and something they're _highly_ unlikely to do anyway

Semi-related, If we want them to use it as a basis, we should probably change it to `from robot import *`?

Adimote replied:
Yes, will fix in separate PR as this feature hasn't been changed in this one.

RealOrangeOne commented on sb/robot/markers.py:
```diff
@@ -0,0 +1,65 @@
+        super(Token, self).__init__(arena)
+        self.marker_id = marker_id
+        self.grabbed = False
+        WIDTH = 0.08
```
This constant should be moved outside the class, or as an attribute of `self`

Adimote replied:
Fixed in https://github.com/sourcebots/simulator/pull/1/commits/ec2ed2f7ea2ca766f20497744dd51afb059efa5d

RealOrangeOne commented on sb/robot/sim_robot.py:
```diff
@@ -36,21 +41,157 @@ def power(self, value):
+                    echo_pin,
+                    direction,
+                ))
+            return 0.0
```
I think rather than returning here, we should re-raise the error. It'll help the students check what pin theyre sending it off

PeterJCLaw replied:
This is complicated -- the real API has no feedback at all about this sort of failure mode.

Much as I see the benefit of reporting the failure here, I _really_ don't like the idea that we're training the competitors that they'll get an error if they use the wrong pins. Doing so leads to a high probability of confusing silent failures on the real robot (which is _much_ worse than a confusing error in the simulator).

Adimote replied:
I suggest maybe adding a warning message? Either way it should _behave_ the same as the real robot IMO.

RealOrangeOne commented on sb/robot/sim_robot.py:
```diff
@@ -156,15 +317,32 @@ def _apply_wheel_force(self, y_position, power):
+            if self.motor_board.m1 == COAST:
+                self._apply_wheel_coast(half_width)
+            else:
+                self._apply_wheel_force(half_width, self.motor_board.m1*100)
```
Spacing around `*` operator, same above

RealOrangeOne commented on sb/robot/vision.py:
```diff
@@ -0,0 +1,20 @@
+                        "id size polar pixel_centre cartesian is_wall_marker is_token_marker")
+
+
+class Marker(MarkerBase):
```
If all this class is doing is adding a `__str__`, I don't think it's necessary

Adimote replied:
Fixed by using the Python 3.5 NamedTuples https://github.com/sourcebots/simulator/pull/1/commits/df207e06ee5baad65b5abb7139c79dbe616407b6

RealOrangeOne reviewed (changes requested):
Few comments, also haven't had the chance to test it yet

PeterJCLaw commented on sb/robot/sim_robot.py:
```diff
@@ -36,21 +41,157 @@ def power(self, value):
+        self._check_voltage(new_voltage)
+        self._motors[1] = new_voltage
+
+    __repr__ = __str__
```
I'm fairly strongly against the idiom of `__repr__ = __str__` since [`__repr__`](https://docs.python.org/3/reference/datamodel.html#object.__repr__) should be a python-ish string which could be used to reconstruct the object. Obviously there is no such constraint on `__str__`.

Even when you want `__str__ = __repr__`, there's pretty much never cause to explicitly say it since that's what the [implementation on `object` does](https://docs.python.org/3/reference/datamodel.html#object.__str__) anyway.

(Further occurrences not noted)

Adimote replied:
Assuming the best method of fixing this is by just removing the`__repr__ = __str__`s, https://github.com/sourcebots/simulator/pull/1/commits/3236aeff0c3142c222bfcc9f5d79ff4843d10fba fixes it.

PeterJCLaw reviewed (commented):
Not sure how much of this code is just being restored or moved (it would have been nice to have had an explicit revert if that was the case).

kierdavis commented on sb/robot/arenas/pirate_islands.py:
```diff
@@ -0,0 +1,157 @@
+                pos = (
+                    0.07 + random.random() * 1.7, 0.07 + random.random() * 1.7)
+                if not positions or min(
+                        [dist(pos, other) for other in positions]) > min_dist:
```
I believe the square brackets can be removed from this line (or possibly they might need to be changed to parentheses instead). This converts the list comprehension into a generator expression, thus avoiding the need to collect the results into a list before passing them to `min`.

Adimote replied:
I didn't know that! fixed in https://github.com/sourcebots/simulator/pull/1/commits/d0936e950c72782eb15ad7f8b7f761e69565ef37

kierdavis commented on sb/robot/arenas/pirate_islands.py:
```diff
@@ -0,0 +1,157 @@
+    @staticmethod
+    def rotate(pos, ang):
+        """
+        Rotate given position around the origin by the given angle
```
I feel this docstring might be clearer if "position" was changed to "vector"

Adimote replied:
fixed in https://github.com/sourcebots/simulator/pull/1/commits/bc84ab12fc7ad7d349676b35677bc351907293eb

kierdavis commented on sb/robot/arenas/pirate_islands.py:
```diff
@@ -0,0 +1,157 @@
+        token_locations = self.random_pos_spaced(4)
+        for zone_id, token_ids in enumerate(
+                [TOKEN_ZONE_0, TOKEN_ZONE_1, TOKEN_ZONE_2, TOKEN_ZONE_3]):
+            for i, (location, token_id) in enumerate(
```
Is `i` actually used in this function?

Adimote replied:
good catch, fixed in https://github.com/sourcebots/simulator/pull/1/commits/cd66fd4f843d017bf81bfcc730b5823d57603805

kierdavis commented on sb/robot/arenas/pirate_islands.py:
```diff
@@ -0,0 +1,157 @@
+                                 start), display.to_pixel_coord(end),
+                             width)
+
+        def line_opposite(start, end, **kwargs):
```
What exactly do "opposite" and "symmetric" mean in this context? Could we rename the functions or add docstrings to clarify what is happening here?

Adimote replied:
After looking at this, it's better to use a rotation system as arenas _should_ be rotationally symmetric. I've made those changes in https://github.com/sourcebots/simulator/pull/1/commits/bc84ab12fc7ad7d349676b35677bc351907293eb

kierdavis commented on sb/robot/game_specific.py:
```diff
@@ -0,0 +1,47 @@
+WALL = set(range(0, 28))  # 0 - 27
```
Almost certainly an issue for a separate PR, but IMO it feels cleaner to import these constants from `robot-api` (which is presumably where they've been copied from) rather than maintaining two copies of this list.

Adimote replied:
Three copies* The ideal situation for this would be to have a game specific repo which every other module uses as a submodule.

Adimote replied:
(this is an issue with the main API however so I don't want to fix it here)

kierdavis commented on sb/robot/sim_robot.py:
```diff
@@ -36,21 +41,157 @@ def power(self, value):
-class Motor:
-    """Represents a motor board."""
-    # This is named `Motor` instead of `MotorBoard` for consistency with pyenv
+BRAKE = 0  # 0 so setting the motors to 0 has exactly the same affect as setting the motors to BRAKE
```
affect -> effect

Adimote replied:
fixed in https://github.com/sourcebots/simulator/pull/1/commits/542b4e532e5edf3cd65f08ce681f2112c2fa2451

kierdavis commented on sb/robot/sim_robot.py:
```diff
@@ -156,15 +317,32 @@ def _apply_wheel_force(self, y_position, power):
                              force_magnitude * sin(self.heading))
         self._body.apply_force(force_world_space, location_world_space)
 
+    def _apply_wheel_coast(self, y_position):
```
Is this just a special case of `_apply_wheel_force` with `power = 0` and a difference coefficient of friction? If so, can we merge the two functions instead of copying them?

Adimote replied:
Fixed in https://github.com/sourcebots/simulator/pull/1/commits/d0f61ebb348445aad93eeffe38638080339bbdd6

Adimote committed:
+ GameObject Getter/Setters for Location & Heading

This means getter and setters which subclasses define don't need
horrible hacks later on. Remove those hacks

Adimote committed:
push __init__ method to top

Adimote committed:
clean up PI arena rotate docstring

Adimote committed:
use generator in min function instead of converting to a list

Adimote committed:
remove redundant enumeration

Adimote committed:
use rotations to lay out lines

Adimote committed:
move constant to class

Adimote committed:
affect->effect

Adimote committed:
remove __repr__ = __str__

Adimote committed:
remove 'NOT SUPPORTED' marker properties

Adimote committed:
add BoardList class for proper indexing of boards

Adimote committed:
remove code duplication in apply_wheel_force

Adimote committed:
define namedtuples with tuples

Adimote committed:
remove pointless `__new__`  call

Adimote committed:
do NamedTuples properly

Adimote committed:
remove unused import

RealOrangeOne commented on sb/robot/arenas/pirate_islands.py:
```diff
@@ -0,0 +1,148 @@
+        super().__init__(arena)
+
+    @property
+    def surface_name(self):
```
This would be so much nicer as simply `surface_name = 'sb/pedestal.png'`?

Adimote replied:
fixed in dbdf7534ccea3a9cf59ebfaa3f5a3319c973ad0b

RealOrangeOne commented on sb/robot/markers.py:
```diff
@@ -0,0 +1,62 @@
+
+    def __init__(self, arena, damping, marker_id=None):
+        self._body = arena._physics_world.create_body(position=(0, 0),
+                                                      angle=0,
```
This indentation style, is really ugly, and not used everywhere in this project, and nowhere in the rest of our codebase. Can you change this over to the standard indenting pattern

Adimote replied:
fixed in 0497121e0216babcca39b6eedf4d77190a7dd641

RealOrangeOne commented:
Reminder that further changes to the API have been made, notably the additional of `spherical` for marker position

Adimote committed:
Don't use NamedTuple as it's python 3.6

We support python 3.5, not 3.6 so we can't use fancy NamedTuple
initialisation

Adimote committed:
remove pointless underscore started variable name

Adimote committed:
convert @property to an actual property

Adimote committed:
fix indentation in markers.py

Adimote commented:
@RealOrangeOne noted, this change is getting quite big so I will make a separate PR to fix the API

Adimote committed:
fix robot not rotating when can is grabbed
